### PR TITLE
Refactor Dropdown to use functional component

### PR DESCRIPTION
**Pull request by sc81**
* Refactor tests

* Refactor DropdownMenu tests to get rid of weird errors

* Add documentation for onToggle and onClose properties

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/dropdown/index.js [hunk 1]
```diff
@@ -6,48 +6,45 @@ import classnames from 'classnames';
...
+		} else if ( onToggle ) {
 			onToggle( isOpen );
 		}
-	}
+		return () => {
+			if ( isOpen && onToggle ) {
+				onToggle( false );
+			}
+		};
+	}, [ isFirstRender, isOpen, onToggle ] );
```
Comment (talldan) [inline packages/components/src/dropdown/index.js:44]: I might be missing an implementation detail, but I think the goal here is to trigger `onToggle` when the state changes, and that could also be handled now as a custom hook, something like the following untested pseudocode:
```jsx
function useObservableState( initialState, onStateChange ) {
    const [ value, setValue ] = useState( initialState );
    return [ value, ( newValue ) => { 
        setValue( newValue ); 
        onStateChange( newValue );
    } ];
}
// ...

const [ isOpen, setIsOpen ] = useIsOpenState( false, onToggle );
```
(I would put this function above the component in the same file)

Would still also need some separate handling for unmounting, but that then becomes simpler:
```jsx
// Fire `onToggle` when the component unmounts
useEffect( () => () => onToggle( false ) );
```

Comment (sc81) [reply packages/components/src/dropdown/index.js:44]: Yes. It should work. I'll check it later.

**Review by talldan — APPROVED** [review]
Thanks for your contribution @sc81. It's great that the docs and tests were also updated as part of this.

I notice that tests are failing at the moment. I think it's worth trying to rebase the PR against the master branch and pushing again, as I think some code that fixes broken tests was merged to master.

**Comment by youknowriad** [comment]
@talldan Should we merge this?

**Comment by talldan** [comment]
Yep, restarted the tests and they look good now 👍

**Comment by sc81** [comment]
No problem. And thank you @talldan

**Pull request merged**